### PR TITLE
[CSL-1608] Forbid isomorphic psks

### PR DESCRIPTION
**Pull request by volhovM**

### CODE REVIEW

File: node/src/Pos/Delegation/Logic/VAR.hs [hunk 1]
```diff
@@ -419,20 +419,42 @@ dlgVerifyBlocks blocks = do
...
+            when (not (isRevokePsk psk) &&
+                  not (stakeholderId `HS.member` richmen)) $
+                throwError $ sformat
+                    ("Block "%build%" contains psk issuers that "%
+                    "don't have enough stake: "%build)
+                    (headerHash blk) psk
+
+            -- Check 4: There are no psks that duplicate previous ones
+            -- compared by (i,d). That is, if i delegated to d using psk1,
+            -- we forbid using psk2 (in the next epoch) if psk2 delegates
```
Comment (int-index) [inline node/src/Pos/Delegation/Logic/VAR.hs]: What does "in the new epoch" mean here? As I said, the issue is reproducible within a single epoch.

**Comment by volhovM** [comment]
@int-index please re-review, i've fixed the problem.

**Comment by int-index** [comment]
And by posting the revocation certificate in the next epoch (rather than the same epoch), all four nodes went into an infinite loop:

```
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] Got request to handle heavyweight psk: ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d }
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] The verdict for cert ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d } is: PHTipMismatch
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] Got request to handle heavyweight psk: ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d }
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] The verdict for cert ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d } is: PHTipMismatch
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] Got request to handle heavyweight psk: ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d }
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] The verdict for cert ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d } is: PHTipMismatch
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] Got request to handle heavyweight psk: ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d }
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] The verdict for cert ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d } is: PHTipMismatch
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] Got request to handle heavyweight psk: ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d }
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] The verdict for cert ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d } is: PHTipMismatch
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] Got request to handle heavyweight psk: ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d }
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] The verdict for cert ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d } is: PHTipMismatch
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] Got request to handle heavyweight psk: ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d }
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] The verdict for cert ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d } is: PHTipMismatch
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] Got request to handle heavyweight psk: ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d }
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] The verdict for cert ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d } is: PHTipMismatch
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] Got request to handle heavyweight psk: ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d }
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] The verdict for cert ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d } is: PHTipMismatch
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] Got request to handle heavyweight psk: ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d }
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] The verdict for cert ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d } is: PHTipMismatch
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] Got request to handle heavyweight psk: ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d }
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] The verdict for cert ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d } is: PHTipMismatch
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] Got request to handle heavyweight psk: ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d }
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] The verdict for cert ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d } is: PHTipMismatch
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] Got request to handle heavyweight psk: ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d }
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] The verdict for cert ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d } is: PHTipMismatch
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] Got request to handle heavyweight psk: ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d }
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] The verdict for cert ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d } is: PHTipMismatch
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] Got request to handle heavyweight psk: ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d }
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] The verdict for cert ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d } is: PHTipMismatch
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] Got request to handle heavyweight psk: ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d }
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] The verdict for cert ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d } is: PHTipMismatch
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] Got request to handle heavyweight psk: ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d }
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] The verdict for cert ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d } is: PHTipMismatch
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] Got request to handle heavyweight psk: ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d }
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] The verdict for cert ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d } is: PHTipMismatch
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] Got request to handle heavyweight psk: ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d }
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] The verdict for cert ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d } is: PHTipMismatch
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] Got request to handle heavyweight psk: ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d }
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] The verdict for cert ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d } is: PHTipMismatch
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] Got request to handle heavyweight psk: ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d }
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] The verdict for cert ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d } is: PHTipMismatch
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] Got request to handle heavyweight psk: ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d }
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] The verdict for cert ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d } is: PHTipMismatch
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] Got request to handle heavyweight psk: ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d }
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] The verdict for cert ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d } is: PHTipMismatch
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] Got request to handle heavyweight psk: ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d }
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] The verdict for cert ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d } is: PHTipMismatch
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] Got request to handle heavyweight psk: ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d }
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] The verdict for cert ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d } is: PHTipMismatch
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] Got request to handle heavyweight psk: ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d }
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] The verdict for cert ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d } is: PHTipMismatch
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] Got request to handle heavyweight psk: ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d }
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] The verdict for cert ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d } is: PHTipMismatch
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] Got request to handle heavyweight psk: ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d }
[node.server.delegation:DEBUG] [2017-09-06 14:17:29 MSK] The verdict for cert ProxySk { w = epoch #0, iPk = pub:2c63dd8d, dPk = pub:2c63dd8d } is: PHTipMismatch
```

**Comment by volhovM** [comment]
Some github bug doesn't show real build status ("in progress" instead of "success") which can be seen on travis website, so I'm merging this.

**Comment by gromakovsky** [comment]
Today in the morning there was the same problem with another PR, I just restarted hlint build and it helped. I've just done it for this PR as well (it's merged already, but anyway).

**Pull request merged**